Stats: Introduce new UTM module

a8ck3n opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #87565 

## Proposed Changes

Introduces a new UTM module behind the "stats/utm-module" feature flag.

<img width="717" alt="SCR-20240219-rvkp" src="https://github.com/Automattic/wp-calypso/assets/40267301/90d77504-198e-4b30-b87b-3e0a789feae4">

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Launch the Calypso Live link.
* Visit the Stats → Traffic page.
* Confirm the UTM module is not visible.
* Add the `stats/utm-module` flag to the URL. — `?flags=stats/utm-module`
* Reload and confirm the new module is present under the Countries module.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

a8ck3n committed:
Add new placeholder module for UTM data.

a8ck3n committed:
Add localized string support.

a8ck3n committed:
Add mock data simulating API request.

a8ck3n committed:
Push data from wrapper to StatsModule.

a8ck3n committed:
Move network request to separate file.

a8ck3n committed:
Put new module behind a feature flag.

grzegorz-cp commented on client/my-sites/stats/stats-module/index.jsx:
```diff
@@ -141,7 +143,15 @@ class StatsModule extends Component {
 
 		// Only show loading indicators when nothing is in state tree, and request in-flight
-		const isLoading = ! this.state.loaded && ! ( data && data.length );
+		let isLoading = ! this.state.loaded && ! ( data && data.length );
```
I don't think that extending this component this way is a good idea. All modules using this file are fetching the data the same way. The new code would create a fork in that logic. This would cascade into having to perform a check in all crucial parts of the component, making two components meshed into one that is complex.

Creating a new component in the `stats-module` directory and using `StatsListCard` with your new fetching mechanism would probably make more sense. You also suspect we'll have to make more changes to reflect the designs; therefore, a new component seems even more appealing.

a8ck3n replied:
That makes a lot of sense. I'll work on those changes today.

a8ck3n replied:
I've introduced a new `StatsModule2` component which is basically just a copy of the existing `StatsModule` component minus the Redux bits (and it's not a class-based component). Was this what you had in mind @grzegorz-cp?

On the one hand, it feels like a lot of code duplication. On the other hand, it could allow us to slowly move the various modules away from the query component set up and over to React Query, at which point we could remove the older component.

Let me know your thoughts on this approach!

grzegorz-cp replied:
Thanks for doing this Kevin. I think this is a good direction, but you are right that there was a lot of duplicated code. I reduced the number of props and renamed it. When we finish extending the component (with all of the custom elements like dropdown and header elements), we'll be able to see if we can extract the duplicated pieces in one shared component.

kangzj commented on client/my-sites/stats/stats-module-utm/useMockData.js:
```diff
@@ -0,0 +1,66 @@
+
+async function fetchMockDataAsync() {
+	// Make an API request. This will fail without an API key but we don't care.
+	const response = await fetch( `https://api.api-ninjas.com/v1/loremipsum?paragraphs=4` );
```
maybe we could skip the data fetch and only stick with the mock data here? It doesn't feel right to have it in the code even it's behind feature flag. What do you think?

a8ck3n replied:
No strong feelings on this from me. I'll pull it out and update the branch.

a8ck3n replied:
Removed the API call in favour of a short timeout to simulate a network request.

grzegorz-cp commented:
CSS changes introduce a gap that is visible when the feature flag is not applied. We'll have to think of a smart way to navigate this. Maybe a class added in JS would be a solution.

a8ck3n committed:
Move UTM module below other modules.

Designs call for this to go under the Countries module but we’ll have to follow up on that as it requires some rethinking of the CSS. Currently we need multiple copies of the Grid CSS depending on which modules are visible.

a8ck3n committed:
Remove fetch call.

Use a simple timeout to simulae a network request. Allows us to see/confirm the loading state is working.

a8ck3n commented:
> CSS changes introduce a gap that is visible when the feature flag is not applied. We'll have to think of a smart way to navigate this. Maybe a class added in JS would be a solution.

I investigated this and it looks like we'll need multiple copies of the grid CSS to account for hidden modules. There are already multiple copies to handle the cases of the Authors and Videos modules potentially being hidden.

For now, I've moved the module below the list. I'm thinking we can place it properly on the page in a follow up PR as it might be worth experimenting with flex here.

a8ck3n committed:
Introduce StatsModule2.

The same as the existing StatsModule but without the Redux bits. Instead, the caller will manage the data and pass it in.

a8ck3n committed:
Restore StatsModule to previous state.

Removed the option to pass in the module state so as not to confuse the logic too much.

a8ck3n committed:
Make sure summary route/page is working.

a8ck3n committed:
Pass summary flag as appropriate.

Prevents the View all/details from showing up on the summary page.

grzegorz-cp committed:
Stats: Reducing the code and colocating similar components.

grzegorz-cp commented:
Thanks for the work @a8ck3n. I updated the PR, colocating the similar components and trimming the file and the props to the crucial ones.
Let's merge this to unblock other tickets and keep iterating.
We also need to extend the controller for Odyssey.

grzegorz-cp committed:
Merge branch 'trunk' into update/stats-introduce-utm-module

grzegorz-cp committed:
Stats: Adding new summary url to the test files.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/11382502

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @a8ck3n for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11382502).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11382502).
